libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp

```diff
@@ -0,0 +1,142 @@
+#define WAI_SET          0XA4
+#define DEVICE_ID        0x66
+#define SENSOR_READY     0x08
+
+
+AP_Airspeed_ASP5033::AP_Airspeed_ASP5033(AP_Airspeed &_frontend, uint8_t _instance) :
+    AP_Airspeed_Backend(_frontend, _instance)
+{
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_ASP5033::init()
+{
+    uint8_t SysID;
+    uint8_t whoami;
+    _dev = hal.i2c_mgr->get_device(get_bus(), ASP5033_I2C_ADDR);
+    if (!_dev) {
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n");
+    }
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+    _dev->write_register(WAI_SET, DEVICE_ID); //Set Whoami ID. 0XA4 is a OPT WhoamI Part ID register, also can be read from address 0x01.
+    //check whoami, set the PartID register to 0X66 first, them Read the PartID register.
+    if (!_dev->read_registers(WAI_REG, &whoami, 1) || //Read the WhoamI ID
+        whoami != DEVICE_ID){
+        _dev->get_semaphore()->give();
+        return false;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor not found\n"); 
+    }
+    _dev->set_retries(2);
+    _dev->get_semaphore()->give();
+    _dev->read_registers(SYS_CONFIG, &SysID, 1); //Read Sensor_config
+    SysID = (SysID&0xFD); //Read the 0xA5 register value, put the read binary value "and" on "11111111101" then write to 0xA5.
+    _dev->write_register(SYS_CONFIG, SysID); //Set Sensor_config
+    _dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_ASP5033::timer, void));
+    return true;
+}         
+
+// read the data from the sensor
+void AP_Airspeed_ASP5033::timer()
+{
+    uint8_t SoC; //Start of conversion flag
+    uint8_t pressraw_bytes[3]; //pressure data AD value
+    uint8_t tempraw_bytes[2]; //temperature data AD value
+    _dev->write_register(CMD, CMD_MEASURE); //once temperature conversion immediately followed by once sensor signal conversion
+    _dev->read_registers(CMD, &SoC, 1); //Judge whether Data collection is over
+    if ((SoC & SENSOR_READY) != 1){   // Read the 0x30 register address. If Sco bit is 0,signify the acquisition end, the data can be read.
+        _dev->read_registers(PRESS_DATA, pressraw_bytes, 3); 
+        //Read ADC output data of pressure
+        //Read 0x06, 0x07, 0x08 register address data to form a 24-bit AD value (pressure data AD value).
+        pressraw = (pressraw_bytes[0]<<24) + (pressraw_bytes[1]<<16) + (pressraw_bytes[2]<<8);
+        //Effective value converting 24 bit integer to 32 bit integer 
+        pressraw >>=15;
+        //Read ADC output data of temperature
+        //Read 0x09, 0x10, register address data to form a 16-bit AD value (temperature data AD value).
+        _dev->read_registers(TEMP_DATA, tempraw_bytes, 2);
+        tempraw = (tempraw_bytes[0]<<8) + tempraw_bytes[1];
+        //Right-shift signed integer back to get correct measurement value
+        tempraw >>=8;
```

Comment:
This doesn't make sense. You are throwing away the lower byte you just carefully added on the line before. If you don't want this then tempraw is just tempraw_bytes[0], and you don't need to do any shiffting/adding.

Reply:
the highest is the symbol bit.

Reply:
My point here is that this would have the same effect as the 2 lines above:
```suggestion
        tempraw = tempraw_bytes[0];
```

You are shifting out the entire tempraw_bytes[1] after you read it. You might as well not bother adding it back in.

Reply:
We can't do that, the highest is the symbol bit. If tempraw = tempraw_bytes[0] that we will lose the symbol bit, then the negative number will be 256 to 0, the temperature could be drop to below zero, so -1 degree become 256 degree, it will get a wrong value when the temperature drop to below zero.